File: flutter_ryft_payment/android/src/main/kotlin/com/ryft/flutter_ryft_payment/FlutterRyftPaymentPlugin.kt
package com.ryft.flutter_ryft_payment

import android.app.Activity
import android.content.Context
import androidx.annotation.NonNull
import com.ryftpay.android.core.model.api.RyftPublicApiKey
import com.ryftpay.android.core.model.payment.*
import com.ryftpay.android.core.service.DefaultRyftPaymentService
import com.ryftpay.android.core.service.RyftPaymentService
import com.ryftpay.android.core.service.listener.RyftPaymentResultListener
import com.ryftpay.android.ui.dropin.*
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import com.ryftpay.android.core.client.RyftApiClient
import com.ryftpay.android.core.client.DefaultRyftApiClientFactory
import com.ryftpay.android.core.model.error.RyftError
import com.ryftpay.android.core.service.listener.RyftLoadPaymentListener

class FlutterRyftPaymentPlugin : FlutterPlugin, MethodCallHandler, ActivityAware, RyftDropInResultListener {
  private lateinit var channel: MethodChannel
  private lateinit var context: Context
  private var activity: Activity? = null
  private var ryftDropIn: RyftDropIn? = null
  private var pendingResult: Result? = null
  private var ryftPublicApiKey: RyftPublicApiKey? = null
  private lateinit var ryftPaymentService: RyftPaymentService
  private var ryftApiClient: RyftApiClient? = null

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_ryft_payment")
    channel.setMethodCallHandler(this)
    context = flutterPluginBinding.applicationContext
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    when (call.method) {
      "initialize" -> {
        val publicApiKey = call.argument<String>("publicApiKey")
        if (publicApiKey == null) {
          result.error("MISSING_ARG", "Public API key is required", null)
          return
        }
        ryftPublicApiKey = RyftPublicApiKey(publicApiKey)
        
        // Initialize the API client and payment service
        ryftApiClient = DefaultRyftApiClientFactory(ryftPublicApiKey!!).createApiClient()
        ryftPaymentService = DefaultRyftPaymentService(ryftApiClient!!)
        
        result.success(null)
      }
      "showDropIn" -> {
        val clientSecret = call.argument<String>("clientSecret")
        val subAccountId = call.argument<String>("subAccountId")
        val collectCardholderName = call.argument<Boolean>("collectCardholderName") ?: false
        val googlePayConfig = call.argument<Map<String, Any>>("googlePayConfig")

        if (clientSecret == null) {
          result.error("MISSING_ARG", "Client secret is required", null)
          return
        }

        val fieldCollection = RyftDropInFieldCollectionConfiguration(
          nameOnCard = collectCardholderName
        )

        var googlePayConfiguration: RyftDropInGooglePayConfiguration? = null
        if (googlePayConfig != null) {
          googlePayConfiguration = RyftDropInGooglePayConfiguration(
            merchantName = googlePayConfig["merchantName"] as String,
            merchantCountryCode = googlePayConfig["merchantCountryCode"] as String
          )
        }

        val configuration = if (subAccountId != null) {
          RyftDropInConfiguration.subAccountPayment(
            clientSecret = clientSecret,
            subAccountId = subAccountId,
            fieldCollection = fieldCollection,
            googlePayConfiguration = googlePayConfiguration,
            publicApiKey = ryftPublicApiKey
          )
        } else {
          RyftDropInConfiguration.standardAccountPayment(
            clientSecret = clientSecret,
            fieldCollection = fieldCollection,
            googlePayConfiguration = googlePayConfiguration,
            publicApiKey = ryftPublicApiKey
          )
        }

        pendingResult = result

        activity?.let {
          ryftDropIn = DefaultRyftDropIn(
            activity = it,
            listener = this
          )
          ryftDropIn?.show(configuration)
        } ?: run {
          result.error("NO_ACTIVITY", "Activity is not available", null)
        }
      }
      "processCardPayment" -> {
        val clientSecret = call.argument<String>("clientSecret")
        val paymentMethodMap = call.argument<Map<String, Any>>("paymentMethod")
        val subAccountId = call.argument<String>("subAccountId")

        if (clientSecret == null || paymentMethodMap == null) {
          result.error("MISSING_ARG", "Client secret and payment method are required", null)
          return
        }

        val cardNumber = paymentMethodMap["cardNumber"] as String
        val expiryMonth = paymentMethodMap["expiryMonth"] as String
        val expiryYear = paymentMethodMap["expiryYear"] as String
        val cvc = paymentMethodMap["cvc"] as String
        val cardholderName = paymentMethodMap["cardholderName"] as? String
        val storeCard = paymentMethodMap["storeCard"] as? Boolean ?: false

        val cardDetails = CardDetails(
          number = cardNumber,
          expiryMonth = expiryMonth,
          expiryYear = expiryYear,
          cvc = cvc,
          nameOnCard = cardholderName
        )

        val options = PaymentMethodOptions.card(store = storeCard)
        val paymentMethod = PaymentMethod.card(cardDetails, options)

        pendingResult = result
        
        // Real implementation using the Ryft SDK
        ryftPaymentService.attemptPayment(
          clientSecret = clientSecret,
          paymentMethod = paymentMethod,
          customerDetails = null,
          subAccountId = subAccountId,
          listener = createPaymentResultListener()
        )
      }
      "processSavedPaymentMethod" -> {
        val clientSecret = call.argument<String>("clientSecret")
        val paymentMethodMap = call.argument<Map<String, Any>>("paymentMethod")
        val subAccountId = call.argument<String>("subAccountId")

        if (clientSecret == null || paymentMethodMap == null) {
          result.error("MISSING_ARG", "Client secret and payment method are required", null)
          return
        }

        val paymentMethodId = paymentMethodMap["id"] as String
        val paymentMethod = PaymentMethod.id(paymentMethodId)

        pendingResult = result
        
        // Real implementation using the Ryft SDK
        ryftPaymentService.attemptPayment(
          clientSecret = clientSecret,
          paymentMethod = paymentMethod,
          customerDetails = null,
          subAccountId = subAccountId,
          listener = createPaymentResultListener()
        )
      }
      "processGooglePayPayment" -> {
        val clientSecret = call.argument<String>("clientSecret")
        val paymentMethodMap = call.argument<Map<String, Any>>("paymentMethod")
        val subAccountId = call.argument<String>("subAccountId")

        if (clientSecret == null || paymentMethodMap == null) {
          result.error("MISSING_ARG", "Client secret and payment method are required", null)
          return
        }

        val token = paymentMethodMap["token"] as String
        val paymentMethod = PaymentMethod.googlePay(token, billingAddress = null)

        pendingResult = result
        
        // Real implementation using the Ryft SDK
        ryftPaymentService.attemptPayment(
          clientSecret = clientSecret,
          paymentMethod = paymentMethod,
          customerDetails = null,
          subAccountId = subAccountId,
          listener = createPaymentResultListener()
        )
      }
      "checkPaymentStatus" -> {
        val paymentSessionId = call.argument<String>("paymentSessionId")
        val clientSecret = call.argument<String>("clientSecret")
        val subAccountId = call.argument<String>("subAccountId")

        if (paymentSessionId == null || clientSecret == null) {
          result.error("MISSING_ARG", "Payment session ID and client secret are required", null)
          return
        }

        pendingResult = result
        
        // Check the payment status
        ryftPaymentService.getLatestPaymentResult(
          paymentSessionId = paymentSessionId,
          clientSecret = clientSecret,
          subAccountId = subAccountId,
          listener = createPaymentResultListener()
        )
      }
      else -> result.notImplemented()
    }
  }
  
  private fun createPaymentResultListener(): RyftPaymentResultListener {
    return object : RyftPaymentResultListener {
      override fun onAttemptingPayment() {
        // Payment is being attempted
      }
      
      override fun onLoadingPaymentResult() {
        // Payment result is being loaded
      }
      
      override fun onPaymentApproved(response: PaymentSession) {
        val result = pendingResult ?: return
        pendingResult = null
        
        val sessionMap = mapOf(
          "id" to response.id,
          "amount" to response.amount,
          "currency" to response.currency.currencyCode,
          "status" to response.status.name
        )
        
        result.success(mapOf(
          "status" to "approved",
          "paymentSession" to sessionMap
        ))
      }
      
      override fun onPaymentRequiresRedirect(returnUrl: String, redirectUrl: String) {
        val result = pendingResult ?: return
        pendingResult = null
        
        result.success(mapOf(
          "status" to "requiresAction",
          "actionType" to "redirect",
          "returnUrl" to returnUrl,
          "redirectUrl" to redirectUrl
        ))
      }
      
      override fun onPaymentRequiresIdentification(returnUrl: String, identifyAction: IdentifyAction) {
        val result = pendingResult ?: return
        pendingResult = null
        
        // In a complete implementation, we'd handle 3DS here
        // For now, we'll return a result indicating that 3DS is required
        result.success(mapOf(
          "status" to "requiresAction",
          "actionType" to "identify",
          "returnUrl" to returnUrl
        ))
      }
      
      override fun onPaymentHasError(lastError: PaymentSessionError) {
        val result = pendingResult ?: return
        pendingResult = null
        
        result.success(mapOf(
          "status" to "failed",
          "errorMessage" to lastError.displayError
        ))
      }
      
      override fun onErrorObtainingPaymentResult(error: RyftError?, throwable: Throwable?) {
        val result = pendingResult ?: return
        pendingResult = null
        
        result.success(mapOf(
          "status" to "failed",
          "errorMessage" to (error?.displayError ?: throwable?.message ?: "Unknown error")
        ))
      }
    }
  }

  override fun onPaymentResult(paymentResult: com.ryftpay.android.ui.model.RyftPaymentResult) {
    val result = pendingResult ?: return
    pendingResult = null

    when (paymentResult) {
      is com.ryftpay.android.ui.model.RyftPaymentResult.Approved -> {
        val sessionMap = mapOf(
          "id" to paymentResult.paymentSession.id,
          "amount" to paymentResult.paymentSession.amount,
          "currency" to paymentResult.paymentSession.currency.currencyCode,
          "status" to paymentResult.paymentSession.status.name
        )

        result.success(mapOf(
          "status" to "approved",
          "paymentSession" to sessionMap
        ))
      }
      is com.ryftpay.android.ui.model.RyftPaymentResult.Failed -> {
        result.success(mapOf(
          "status" to "failed",
          "errorMessage" to paymentResult.error.displayError
        ))
      }
      is com.ryftpay.android.ui.model.RyftPaymentResult.Cancelled -> {
        result.success(mapOf(
          "status" to "cancelled"
        ))
      }
    }
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    activity = binding.activity
  }

  override fun onDetachedFromActivityForConfigChanges() {
    activity = null
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    activity = binding.activity
  }

  override fun onDetachedFromActivity() {
    activity = null
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
} 

File: flutter_ryft_payment/ios/Classes/FlutterRyftPaymentPlugin.swift
import Flutter
import UIKit
// In a real implementation, uncomment these imports
import RyftCore
import RyftUI

public class FlutterRyftPaymentPlugin: NSObject, FlutterPlugin, RyftDropInPaymentDelegate {
  private var pendingResult: FlutterResult?
  // In a real implementation, this would be RyftDropInPaymentViewController
  private var ryftDropIn: RyftDropInPaymentViewController?
  private var publicApiKey: String?
  private var apiClient: RyftApiClient?
  
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "flutter_ryft_payment", binaryMessenger: registrar.messenger())
    let instance = FlutterRyftPaymentPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }
  
  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "initialize":
      if let args = call.arguments as? [String: Any],
         let publicApiKey = args["publicApiKey"] as? String {
        self.publicApiKey = publicApiKey
        
        // Initialize the API client
        let environment = RyftCore.RyftEnvironment.fromPublicKey(publicApiKey)
        self.apiClient = DefaultRyftApiClient(publicApiKey: publicApiKey, environment: environment)
        
        result(nil)
      } else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments", details: nil))
      }
      
    case "showDropIn":
      guard let args = call.arguments as? [String: Any],
            let clientSecret = args["clientSecret"] as? String,
            let publicApiKey = self.publicApiKey else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing required arguments", details: nil))
        return
      }
      
      let subAccountId = args["subAccountId"] as? String
      let collectCardholderName = args["collectCardholderName"] as? Bool ?? false
      
      var applePayConfig: RyftApplePayConfig?
      if let applePayArgs = args["applePayConfig"] as? [String: Any],
         let merchantIdentifier = applePayArgs["merchantIdentifier"] as? String,
         let merchantName = applePayArgs["merchantName"] as? String,
         let merchantCountryCode = applePayArgs["merchantCountryCode"] as? String {
        applePayConfig = RyftApplePayConfig(
          merchantIdentifier: merchantIdentifier,
          merchantCountryCode: merchantCountryCode,
          merchantName: merchantName
        )
      }
      
      let fieldCollection = RyftDropInConfiguration.RyftDropInFieldCollectionConfig(
        nameOnCard: collectCardholderName
      )
      
      let config = RyftDropInConfiguration(
        clientSecret: clientSecret,
        accountId: subAccountId,
        applePay: applePayConfig,
        fieldCollection: fieldCollection
      )
      
      pendingResult = result
      
      if let viewController = UIApplication.shared.keyWindow?.rootViewController {
        ryftDropIn = RyftDropInPaymentViewController(
          config: config,
          publicApiKey: publicApiKey,
          delegate: self
        )
        viewController.present(ryftDropIn!, animated: true, completion: nil)
      } else {
        result(FlutterError(code: "NO_VIEW_CONTROLLER", message: "No view controller available", details: nil))
      }
      
    case "processCardPayment":
      guard let args = call.arguments as? [String: Any],
            let clientSecret = args["clientSecret"] as? String,
            let paymentMethodMap = args["paymentMethod"] as? [String: Any],
            let apiClient = self.apiClient else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing required arguments", details: nil))
        return
      }
      
      let subAccountId = args["subAccountId"] as? String
      
      // Extract card details
      let cardNumber = paymentMethodMap["cardNumber"] as! String
      let expiryMonth = paymentMethodMap["expiryMonth"] as! String
      let expiryYear = paymentMethodMap["expiryYear"] as! String
      let cvc = paymentMethodMap["cvc"] as! String
      let cardholderName = paymentMethodMap["cardholderName"] as? String
      let storeCard = paymentMethodMap["storeCard"] as? Bool ?? false
      
      // Create card request
      let card = RyftCore.CardRequest(
        number: cardNumber,
        expiryMonth: expiryMonth,
        expiryYear: expiryYear,
        cvc: cvc,
        nameOnCard: cardholderName
      )
      
      pendingResult = result
      
      // Process payment with card
      apiClient.attemptPayment(
        request: AttemptPaymentRequest.fromCard(
          clientSecret: clientSecret,
          card: card,
          storeCard: storeCard
        ),
        accountId: subAccountId,
        completion: { [weak self] paymentResult in
          self?.handleApiPaymentResult(paymentResult, result: result)
        }
      )
      
    case "processSavedPaymentMethod":
      guard let args = call.arguments as? [String: Any],
            let clientSecret = args["clientSecret"] as? String,
            let paymentMethodMap = args["paymentMethod"] as? [String: Any],
            let apiClient = self.apiClient else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing required arguments", details: nil))
        return
      }
      
      let subAccountId = args["subAccountId"] as? String
      let paymentMethodId = paymentMethodMap["id"] as! String
      
      pendingResult = result
      
      // Process payment with saved payment method
      apiClient.attemptPayment(
        request: AttemptPaymentRequest.fromPaymentMethod(
          clientSecret: clientSecret,
          paymentMethodId: paymentMethodId
        ),
        accountId: subAccountId,
        completion: { [weak self] paymentResult in
          self?.handleApiPaymentResult(paymentResult, result: result)
        }
      )
      
    case "processApplePayPayment":
      guard let args = call.arguments as? [String: Any],
            let clientSecret = args["clientSecret"] as? String,
            let publicApiKey = self.publicApiKey else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing required arguments", details: nil))
        return
      }
      
      let subAccountId = args["subAccountId"] as? String
      
      // For Apple Pay, we'd usually create a RyftApplePayComponent and handle the payment
      // Here we'll use a simplified approach that directly calls the API
      
      // Note: In a real implementation, this would be more complex and involve
      // presenting the Apple Pay sheet to the user
      
      pendingResult = result
      
      // For now, just notify that Apple Pay requires interactive handling
      result([
        "status": "failed",
        "errorMessage": "Apple Pay requires user interaction. Please use showDropIn instead."
      ])
      
    case "checkPaymentStatus":
      guard let args = call.arguments as? [String: Any],
            let paymentSessionId = args["paymentSessionId"] as? String,
            let clientSecret = args["clientSecret"] as? String,
            let apiClient = self.apiClient else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing required arguments", details: nil))
        return
      }
      
      let subAccountId = args["subAccountId"] as? String
      
      pendingResult = result
      
      // Get the latest payment session
      apiClient.getPaymentSession(
        id: paymentSessionId,
        clientSecret: clientSecret,
        accountId: subAccountId,
        completion: { [weak self] paymentResult in
          self?.handleApiPaymentResult(paymentResult, result: result)
        }
      )
      
    default:
      result(FlutterMethodNotImplemented)
    }
  }
  
  private func handleApiPaymentResult(_ paymentResult: Result<PaymentSession, HttpError>, result: FlutterResult) {
    switch paymentResult {
    case .success(let paymentSession):
      if let requiredAction = paymentSession.requiredAction {
        // Payment requires additional action
        result([
          "status": "requiresAction",
          "actionType": requiredAction.type.rawValue,
          "paymentSession": [
            "id": paymentSession.id,
            "amount": paymentSession.amount,
            "currency": paymentSession.currency,
            "status": paymentSession.status.rawValue
          ]
        ])
      } else {
        // Payment successful
        result([
          "status": "approved",
          "paymentSession": [
            "id": paymentSession.id,
            "amount": paymentSession.amount,
            "currency": paymentSession.currency,
            "status": paymentSession.status.rawValue
          ]
        ])
      }
      
    case .failure(let error):
      // Payment failed
      result([
        "status": "failed",
        "errorMessage": error.localizedDescription
      ])
    }
  }
  
  // MARK: - RyftDropInPaymentDelegate
  public func onPaymentResult(result: RyftPaymentResult) {
    guard let pendingResult = pendingResult else { return }
    self.pendingResult = nil
    
    switch result {
    case .success(let paymentSession):
      let sessionMap: [String: Any] = [
        "id": paymentSession.id,
        "amount": paymentSession.amount,
        "currency": paymentSession.currency,
        "status": paymentSession.status.rawValue
      ]
      
      pendingResult([
        "status": "approved",
        "paymentSession": sessionMap
      ])
      
    case .failed(let error):
      pendingResult([
        "status": "failed",
        "errorMessage": error.displayError
      ])
      
    case .cancelled:
      pendingResult([
        "status": "cancelled"
      ])
      
    case .pendingAction(let paymentSession, let requiredAction):
      // Handle 3DS actions
      ryftDropIn?.handleRequiredAction(
        returnUrl: URL(string: paymentSession.returnUrl),
        requiredAction
      )
    }
  }
} 